File: iOS/MonitorApp/ObjC/MonitorApp/CustomerTableViewCell.m
//
//  CustomerTableViewCell.m
//  MonitorApp
//
//  Created by Radwar on 8/24/17.
//  Copyright © 2017 curbside. All rights reserved.
//

#import "CustomerTableViewCell.h"
#import "MonitorExtras.h"
@import Curbside;

@interface CustomerTableViewCell ()

@property (strong, nonatomic) IBOutlet UILabel *customerNameTIDLabel;
@property (strong, nonatomic) IBOutlet UILabel *etaValueLabel;
@property (strong, nonatomic) IBOutlet UILabel *distanceValueLabel;
@property (strong, nonatomic) IBOutlet UILabel *dateValueLabel;
@property (strong, nonatomic) IBOutlet UILabel *statusLabel;
@property (strong, nonatomic) IBOutlet UILabel *trackTokenLabel;

@property (strong, nonatomic) IBOutlet UIButton *endTripButton;
@property (strong, nonatomic) IBOutlet UIButton *cancelTripButton;

@property (strong, nonatomic) NSString *trackToken;
@end

@implementation CustomerTableViewCell

- (void)awakeFromNib {
    [super awakeFromNib];
    _endTripButton.layer.cornerRadius = 4.0;
    _cancelTripButton.layer.cornerRadius = 4.0;
}

- (void)setSelected:(BOOL)selected animated:(BOOL)animated {
    [super setSelected:selected animated:animated];
}

- (void)setUserStatusUpdate:(CSUserStatusUpdate *)userStatusUpdate
{
    _userStatusUpdate = userStatusUpdate;
    NSString *customerName = userStatusUpdate.userInfo.fullName;
    if (customerName.length == 0)
        customerName = userStatusUpdate.trackingIdentifier;
    _customerNameTIDLabel.text = customerName;
    int distanceFromSite = userStatusUpdate.distanceFromSite;
    _distanceValueLabel.text = monitor_FormatDistance(userStatusUpdate.distanceFromSite);
    int eta = userStatusUpdate.estimatedTimeOfArrival;
    _etaValueLabel.text = eta > 0 ?  monitor_FormatSeconds(eta) : @"-";
    _dateValueLabel.text = FormatDate(userStatusUpdate.lastUpdateTimestamp);
    switch (_userStatusUpdate.userStatus) {
        case CSUserStatusArrived:
            _statusLabel.text = @"Arrived";
            break;
        case CSUserStatusInTransit:
            _statusLabel.text = @"In Transit";
            break;
        case CSUserStatusApproaching:
            _statusLabel.text = @"Approaching Site";
            break;
        case CSUserStatusUserInitiatedArrived:
            _statusLabel.text = @"Customer Initiated Arrived";
            break;
        case CSUserStatusUnknown:
            _statusLabel.text = @"Unknown";
            if (distanceFromSite == 0)
                _distanceValueLabel.text = @"-";
            break;
    }
    
#warning Note this is taking the first trackToken and using that trackToken for dispaly and cancel/Stop.
    CSTripInfo *firstTripInfo = [_userStatusUpdate.tripInfos firstObject];
    _trackToken = firstTripInfo.trackToken;
    _trackTokenLabel.text = _trackToken;
}

- (IBAction)cancelTrip:(id)sender {
    [[CSMonitoringSession currentSession] cancelTripForTrackingIdentifier:_userStatusUpdate.trackingIdentifier trackTokens:@[_trackToken]];
}

- (IBAction)endTrip:(id)sender {
    [[CSMonitoringSession currentSession] completeTripForTrackingIdentifier:_userStatusUpdate.trackingIdentifier trackTokens:@[_trackToken]];
}
@end

File: iOS/Swift/MonitorApp/MonitorApp/CustomerTableViewCell.swift
//
//  CustomerTableViewCell.swift
//  MonitorApp
//
//  Created by Hon Chen on 9/7/17.
//  Copyright © 2017 curbside. All rights reserved.
//

import UIKit
import Curbside

let kCustomerTableViewCellIdentifier = "kCustomerTableViewCellIdentifier"

class CustomerTableViewCell: UITableViewCell {

    var userStatusUpdate: CSUserStatusUpdate? {
        didSet {
            updateUI()
        }
    }
    
    @IBOutlet var customerNameTIDLabel: UILabel!
    @IBOutlet var etaValueLabel: UILabel!
    @IBOutlet var distanceValueLabel: UILabel!
    @IBOutlet var dateValueLabel: UILabel!
    @IBOutlet var statusLabel: UILabel!
    
    @IBOutlet var endTripButton: UIButton!
    @IBOutlet var cancelTripButton: UIButton!
    
    override func awakeFromNib() {
        super.awakeFromNib()
        endTripButton.layer.cornerRadius = 4.0
        cancelTripButton.layer.cornerRadius = 4.0
    }
    
    private func updateUI() {
        guard let userStatusUpdate = userStatusUpdate else {
            customerNameTIDLabel.text = nil
            distanceValueLabel.text = nil
            etaValueLabel.text = nil
            dateValueLabel.text = nil
            statusLabel.text = nil
            return
        }
        
        // set customer name TID
        if let customerName = userStatusUpdate.userInfo?.fullName, !(customerName.characters.isEmpty) {
            customerNameTIDLabel.text = customerName
        } else {
            customerNameTIDLabel.text = userStatusUpdate.trackingIdentifier
        }
        
        // set distance
        distanceValueLabel.text = FormatDistance(Float(userStatusUpdate.distanceFromSite))
        
        // set eta
        let eta = userStatusUpdate.estimatedTimeOfArrival
        etaValueLabel.text = eta > 0 ? FormatSeconds(eta) : "-"
        
        // set date
        dateValueLabel.text = FormatDate(userStatusUpdate.lastUpdateTimestamp)
        
        // set user status
        switch userStatusUpdate.userStatus {
        case .arrived:
            statusLabel.text = "Arrived"
        case .inTransit:
            statusLabel.text = "In Transit"
        case .approaching:
            statusLabel.text = "Approaching Site"
        case .userInitiatedArrived:
            statusLabel.text = "Customer Initiated Arrived"
        case .unknown:
            statusLabel.text = "Unknown"
        }
    }
    
    @IBAction func endTrip(_ sender: Any) {
        if let trackingIdentifier = userStatusUpdate?.trackingIdentifier {
            CSMonitoringSession.current().completeTrip(forTrackingIdentifier: trackingIdentifier, trackTokens: nil)
        }
    }
    
    @IBAction func cancelTrip(_ sender: Any) {
        if let trackingIdentifier = userStatusUpdate?.trackingIdentifier {
            CSMonitoringSession.current().cancelTrip(forTrackingIdentifier: trackingIdentifier, trackTokens: nil)
        }
    }

}
